sage.py

```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-    #        print(sample)
-    #        print('--', s)
-    print('Total medium states', len(med_states))  
-    print('Total severe states', len(sev_states))
-    return(med_states, sev_states, sev_sinks)
-
-## collecting sub-behaviors back into the same trace -- condensed_data is the new object to deal with
-def make_condensed_data(alerts, keys, state_traces, med_states, sev_states):
-    levelone = set()
-    levelone_ben = set()
-    condensed_data = dict()
+        med_sev = [int(state) for sym, state in zip(sample, state_list) if len(str(rev_smallmapping[sym.split('|')[0]])) == 2]
+        med_sev_states.update(med_sev)
+        high_sev = [int(state) for sym, state in zip(sample, state_list) if len(str(rev_smallmapping[sym.split('|')[0]])) == 3]
+        high_sev_states.update(high_sev)
+        sev_sinks.update(_sev_sinks)
+
+    print('Traces in sinks:', traces_in_sinks, 'Total traces:', total_traces, 'Percentage:', 100 * (traces_in_sinks / float(total_traces)))
+    print('Total medium-severity states:', len(med_sev_states))
+    print('Total high-severity states:', len(high_sev_states))
+    print('Total severe sinks:', len(sev_sinks))
+    return state_traces, med_sev_states, high_sev_states, sev_sinks
+
+
+# Step 6.3: Create state sequences (collecting sub-behaviors back into the same trace, augmented with state IDs)
+def make_state_sequences(episode_subsequences, state_traces, med_states, sev_states):
+    level_one = set()  # TODO: what to do with these states?
+    state_sequences = dict()
     counter = -1
-    for tid, (attacker, episodes) in enumerate(zip(keys, alerts)):
+    for tid, (attack, episode_subsequence) in enumerate(episode_subsequences.items()):
 
-        if len(episodes) < 3:
+        if len(episode_subsequence) < 3:
             continue
-        #print(' ------------- COUNTER ', counter, '------')
         counter += 1
         
-        if '10.0.254' not in attacker:
+        if '10.0.254' not in attack:  # TODO: this part has to be commented for CCDC dataset
             continue
-        if ('147.75' in attacker or '69.172'  in attacker):
-                continue
-        tr = [int(x) for x in state_traces[counter]]
-        #print(counter)
-        num_servs = [len(set((x[6]))) for x in episodes]
-        max_servs = [most_frequent(x[6]) for x in episodes]
-        #print(max_servs)
-        
-        if 0 in tr and (not set(tr).isdisjoint(sev_states) or not set(tr).isdisjoint(med_states)):
-            levelone.add(tr[tr.index(0)+1])
-       
-        #print([x[2] for x in episodes]) 
-        #print(state_traces[counter])
-        new_state = (state_traces[counter][1:])[::-1]
+        if '147.75' in attack or '69.172' in attack:
+            continue
+
+        trace = [int(state) for state in state_traces[counter]]
+        max_services = [_most_frequent(epi[6]) for epi in episode_subsequence]
+
+        if 0 in trace and (not set(trace).isdisjoint(sev_states) or not set(trace).isdisjoint(med_states)):
```

Comment:
I don't really understand the `level_one` here. All traces (unless `0` is removed above) start with `0` (the root node). And the if-check: if this trace has med- or high-sev nodes? Below, the print statement says "High-severity objective states", but is it the case? And those objective states are calculated when creating AGs...

Reply:
This part can be removed. Probably was being used for testing at some point.